### PR TITLE
[build] remove autotools

### PR DESCRIPTION
**Pull request by bukepo**
This commit removes autotools support from OpenThread project.

* removed autotool tests.
* migrate tests depending on autotool to cmake

### CODE REVIEW

**Comment by bukepo** [comment]
submitting this PR as draft to see if anyone wants to keep the nlbuild autotool support.

**Comment by wgtdkp** [comment]
> submitting this PR as draft to see if anyone needs the nlbuild autotool support.

+1 on getting rid of the ancient autotools!

File: .github/workflows/makefile-check.yml [hunk 1]
```diff
@@ -1,60 +0,0 @@
...
-      uses: step-security/harden-runner@6b3083af2869dc3314a0257a42f4af696cc79ba3 # v2.3.1
-      with:
-        egress-policy: audit # TODO: change to 'egress-policy: block' after couple of runs
-
-    - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
-      with:
-        submodules: true
-    - name: Check
-      run: |
-        script/check-core-makefiles
```
Comment (jwhui) [inline .github/workflows/makefile-check.yml:60]: We probably want to keep this, as it also checks CMakeLists.txt and BUILD.gn files.

File: .github/workflows/toranj.yml [hunk 2]
```diff
@@ -71,9 +71,10 @@ jobs:
         GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
       run: |
         sudo rm /etc/apt/sources.list.d/* && sudo apt-get update
+        sudo apt-get --no-install-recommends install -y ninja-build lcov cmake
     - name: Build & Run
       run: |
-        top_builddir=$(pwd)/build/toranj ./tests/toranj/start.sh
+        TORANJ_VERBOSE=${RUNNER_DEBUG-0} top_builddir=$(pwd)/build/toranj ./tests/toranj/start.sh
```
Comment (abtink) [inline .github/workflows/toranj.yml:77]: Is this disabled by default? How do we enable `RUNNER_DEBUG-0` mode?

File: script/check-posix-build-autotools [hunk 3]
```diff
@@ -1,64 +0,0 @@
...
-
-        reset_source
-        ./tests/toranj/build.sh posix
-
-        reset_source
-        ./tests/toranj/build.sh posix-15.4+trel
-
-        reset_source
-        ./tests/toranj/build.sh posix-trel
-    fi
```
Comment (abtink) [inline script/check-posix-build-autotools:54]: Do we cover the same set (posix build with different radio combo) somewhere now?

File: script/check-simulation-build-autotools [hunk 4]
```diff
@@ -1,278 +0,0 @@
...
-        "-DOPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_PLATFORM_USEC_TIMER_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_SNTP_CLIENT_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_SRP_CLIENT_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_TMF_NETDIAG_CLIENT_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_UDP_FORWARD_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_MAC_BEACON_PAYLOAD_PARSING_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_MAC_OUTGOING_BEACON_PAYLOAD_ENABLE=1"
```
Comment (abtink) [inline script/check-simulation-build-autotools:92]: Do we cover the same set of configs under `cmake` now (e.g. in `build_all_features()`.  e.g. do we have `OT_TX_BEACON_PAYLOAD`  and other configs here?

It is important to cover all features in `build_all_features()` so to make sure a new PR would not break any component.

File: script/check-simulation-build-autotools [hunk 5]
```diff
@@ -1,278 +0,0 @@
...
-        "-DOPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_DUA_ENABLE=1"
-        "-DOPENTHREAD_CONFIG_MLR_ENABLE=1"
-    )
-
-    # Build Thread 1.1 with full features and no log
-    export CPPFLAGS="${options[*]}"
-    reset_source
-    make -f examples/Makefile-simulation THREAD_VERSION=1.1 LOG_OUTPUT=NONE
```
Comment (abtink) [inline script/check-simulation-build-autotools:107]: Do we cover this case (no-logs) under cmake?

File: script/check-simulation-build-autotools [hunk 6]
```diff
@@ -1,278 +0,0 @@
...
-
-    # Build Thread 1.3 with full features and full logs
-    export CPPFLAGS="${options[*]} ${options_1_3[*]}"
-    reset_source
-    make -f examples/Makefile-simulation THREAD_VERSION=1.3 FULL_LOGS=1
-
-    # Build Thread 1.1 with ASSERT disabled
-    export CPPFLAGS="${options[*]} -DOPENTHREAD_CONFIG_ASSERT_ENABLE=0"
-    reset_source
-    make -f examples/Makefile-simulation THREAD_VERSION=1.1
```
Comment (abtink) [inline script/check-simulation-build-autotools:127]: Same with all of these?

File: script/check-simulation-build-autotools [hunk 7]
```diff
@@ -1,278 +0,0 @@
...
-        --with-examples=simulation \
-        --disable-docs \
-        --disable-tests \
-        --with-vendor-extension=../src/core/common/extension_example.cpp \
-        --with-ncp-vendor-hook-source=../src/ncp/example_vendor_hook.cpp
-    make -j"${OT_BUILD_JOBS}"
-    cd ..
-}
-
-build_multi_radio_links()
```
Comment (abtink) [inline script/check-simulation-build-autotools:212]: Do we cover something similar under cmake scripts?

**Comment by bukepo** [comment]
@abtink could you please take look at the toranj changes? I increased the timeout as I found `Toranj / toranj-ncp-15.4`  consistently failed without the extra delay. Also if ran locally, it always passed. If I ran on GitHub Action in verbose mode, it passed, so I think there is some race condition when wpantund starts causing it not responding to wpanctl requests. And the extra delay can somehow bypass it.

File: tests/toranj/build.sh [hunk 8]
```diff
@@ -206,129 +187,126 @@ case ${build_config} in
...
 
     cli-15.4+trel | cli-trel+15.4)
         echo "==================================================================================================="
         echo "Building OpenThread NCP FTD mode with simulation platform - multi radio (15.4 + TREL)"
         echo "==================================================================================================="
         cd "${top_builddir}" || die "cd failed"
-        cmake -GNinja -DOT_PLATFORM=simulation -DOT_COMPILE_WARNING_AS_ERROR=ON -DOT_COVERAGE=${ot_coverage} \
+        cmake -GNinja -DOT_PLATFORM=simulation -DOT_COMPILE_WARNING_AS_ERROR=ON -DOT_COVERAGE=${coverage} \
+            -DOT_DOC=OFF "-DBUILD_TESTING=$tests" -DOT_LOG_LEVEL_DYNAMIC=ON \
+            -DOT_MAC_FILTER=ON \
```
Comment (abtink) [inline tests/toranj/build.sh:224]: Why adding the flags `-DOT_LOG_LEVEL_DYNAMIC=ON -DOT_MAC_FILTER=ON` here?

We have the `openthead-core-toran-configj.h` header file which defines all configs for toranj tests. It already defines both of these.

Comment (abtink) [reply tests/toranj/build.sh:224]: I see the same below in all other changes, should remove them from all?

File: .github/workflows/toranj.yml (hunk 2 above)
Comment (abtink) [reply .github/workflows/toranj.yml:77]: Suggest removing this.

Comment (bukepo) [reply .github/workflows/toranj.yml:77]: Yes, it's disabled by default. We can enable debug mode when re-run a test, which can be useful. If you still prefer removing this, I can remove it.

**Comment by abtink** [comment]
> @abtink could you please take look at the toranj changes? I increased the timeout as I found Toranj / toranj-ncp-15.4 consistently failed without the extra delay. Also if ran locally, it always passed. If I ran on GitHub Action in verbose mode, it passed, so I think there is some race condition when wpantund starts causing it not responding to wpanctl requests. And the extra delay can somehow bypass it.

@bukepo suggest keeping the tests as they are and limiting the change in this PR to only build scripts.

File: script/check-posix-build-autotools (hunk 3 above)
Comment (bukepo) [reply script/check-posix-build-autotools:54]: I assumed `check-posix-build-autotools` and `check-posix-build-cmake` were equivalent. I'll update the scripts.

File: .github/workflows/toranj.yml (hunk 2 above)
Comment (abtink) [reply .github/workflows/toranj.yml:77]: I suggest we limit the changes in this PR to its main goal of removing autotool makefiles.
We can add other enhancements (like this) in follow-up seperate PRs?

File: tests/toranj/build.sh (hunk 8 above)
Comment (abtink) [reply tests/toranj/build.sh:224]: Is this addresses in new push? Looks like it same as before?

**Comment by abtink** [comment]
@bukepo  I think it may be better to break this into smaller changes. 
To help with this I just submitted https://github.com/openthread/openthread/pull/9115 which updates `toranj` build scripts to use CMake.

File: tests/toranj/build.sh (hunk 8 above)
Comment (bukepo) [reply tests/toranj/build.sh:224]: No, I'm still working on this PR. I'll mark it as draft before all tests pass again.

**Comment by bukepo** [comment]
> @bukepo I think it may be better to break this into smaller changes. To help with this I just submitted #9115 which updates `toranj` build scripts to use CMake.

Thanks! That will be great helpful!

File: tests/toranj/build.sh (hunk 8 above)
Comment (abtink) [reply tests/toranj/build.sh:224]: Sounds good. I mark these as resolved.

**Review by abtink — APPROVED** [review]
LGTM. Thanks @bukepo for helping with this. 👍

**Review by jwhui — APPROVED** [review]
Thanks! 👍🏼

**Pull request merged**